feat(rest): enable Express middleware to be used as interceptors with LoopBack

raymondfeng opened the pull request:
Depends on https://github.com/strongloop/loopback-next/pull/5317
Implements #2035 
Supersedes #2434  

This PR adds ability to use Express middleware with LoopBack 4 as follows:

1. The default sequence now has `Middleware` step. It creates an invocation chain to call registered middleware handlers via extension point/extension pattern.
2. The sequence can be customized to have more than one `Middleware` steps.
3. Express middleware can also be wrapped as LB4 interceptors, which can in turn be added to global/class/method level.
4. Move built-in cors and openapi endpoints to middleware action.

restServer -> sequence (middleware, findRoute, ..., invoke, send) --> global middleware interceptor --> class/method level interceptor.

Some of the ideas are inspired by https://github.com/codejamninja/lb4-middleware. Thanks @codejamninja and @victor-enogwe

See docs at:

- https://github.com/strongloop/loopback-next/blob/middleware-interceptor/docs/site/Middleware.md.
- https://github.com/strongloop/loopback-next/blob/middleware-interceptor/docs/site/Express-middleware.md

![middleware](https://raw.githubusercontent.com/strongloop/loopback-next/middleware-interceptor/docs/site/imgs/middleware.png)

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

raymondfeng commented:
See tests at https://github.com/strongloop/loopback-next/tree/middleware-interceptor/packages/rest/src/__tests__/acceptance/middleware.

hacksparrow commented on docs/site/Express-middleware.md:
````diff
@@ -0,0 +1,229 @@
+Now we can set up applications to leverage the new sequence:
+
+```ts
+// Create another middleware phase
````
Which file is this? `src/index.ts`?

hacksparrow commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,229 @@
+  .bind('middleware.afterInvoke')
+  .toProvider(MiddlewareProvider)
+  .tag({[CoreTags.EXTENSION_POINT]: POST_INVOCATION_MIDDLEWARE});
+helper.app.sequence(SequenceWithMiddleware);
```
What is `helper`? Where does it come from? Can't we just use `app.sequence(SequenceWithMiddleware)`?

hacksparrow commented on docs/site/Express-middleware.md:
`````diff
@@ -0,0 +1,229 @@
+````
+
+```ts
+const binding = app.middleware(spy, spyConfig, {
`````
Would `app.middleware(mw, mwConfig, {` or `app.middleware(middleware, middleware Config, {` be more descriptive?

hacksparrow replied:
Or include the definition of `spy` and `spyConfig` in the code.

hacksparrow commented on docs/site/Express-middleware.md:
`````diff
@@ -0,0 +1,229 @@
+
+#### Register an Express middleware for the sequence
+
+````ts
`````
Which file is this? Will users have to code this? If no, then we should have add some context to it.

hacksparrow commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,229 @@
+With the ability to wrap Express middleware as LoopBack 4 interceptors, we can
+use the same programming model to register middleware as global interceptors or
+local interceptors denoted by `@intercept` decorators at class and method
+levels.
```
Need an example showing how to do it.

hacksparrow commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,229 @@
+### Introduce middleware to REST sequence of actions
+
+There are a few steps involved in the default sequence of actions. See
+`Sequence.md` for more details.
```
Will a html link be generated from `Sequence.md`?

raymondfeng replied:
yes.

hacksparrow commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,229 @@
+We create a few helper functions that wraps an Express middleware module into an
+LoopBack 4 interceptor.
+
+### Create an interceptor from Express middleware module name or factory function
```
Since this file will be a reference doc, we should prioritize on user API documentation. Including implementation details should be an added bonus, we could even omit it, IMO.

raymondfeng replied:
We can improve it once the apidocs is live.

bajtos commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,229 @@
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      if (this.middleware != null) {
```
Instead of using an `if` block here, can you set `middleware` to a no-op function in the constructor when no value is provided by dependency injection?

bajtos commented on packages/rest/package.json:
```diff
@@ -44,6 +44,7 @@
     "lodash": "^4.17.15",
     "on-finished": "^2.3.0",
     "path-to-regexp": "^6.1.0",
+    "p-event": "^4.1.0",
```
Do we need to add `@types/p-event` to our dev dependencies?

raymondfeng replied:
No, `p-event@4.1.0` has self-contained `d.ts`.

bajtos commented on packages/rest/src/middleware/middleware-interceptor.ts:
```diff
@@ -0,0 +1,221 @@
+): GenericInterceptor<CTX> {
+  let factory: MiddlewareFactory<CFG>;
+  if (typeof middleware === 'string') {
+    factory = require(middleware);
```
Please note this `require` call will resolve the path relatively to the location of `@loopback/rest` module, not relatively to the code calling `createInterceptor`. This creates confusing behavior when `createInterceptor` is called e.g. from a 3rd-party package that's symlinked into `node_modules` tree.

We have been bitten by this problem multiple times in the past. While there are ways how to work around it, I feel the complexity is not worth the benefits.

Can you please remove the string-based variant and always require the factory function instead?

Also what if the middleware does not provide any factory function and all the users have is a regular function `(req, res, next) => void`? Are they supposed to wrap that function into a factory just for the sake of satisfying this API?

Last but not least, how do you envision caching the instances of middleware between requests? It's important to call middleware factory function only once at app startup, we must not call it for every incoming request. I don't yet fully comprehend how is `createInterceptor` going to be used, so maybe this is not a problem? IDK.

I also feel this function is violating Single Responsibility Principle, because it has two responsibilities:
- Obtain an instance of middleware handler.
- Convert a middleware handler to an interceptor.

Have you considering removing the first responsibility and modifying the function to accept the middleware handler only?

raymondfeng replied:
- Support for middleware module as `string` is removed.
- Default `InvokeMiddleware` is set to no-op.
- Caching is handled at binding scope level.

bajtos commented:
@raymondfeng this is a sizeable change with major design implications, can you please work with @dhmlau to create space in our plan to give us all enough time to review the proposal?

raymondfeng commented:
FYI: I did further refactoring:

1. Added `@loopback/middleware` as the infrastructure to allow Koa style middleware and make it easy to plug in existing Express middleware as middleware or interceptors so that they can be used at sequence/action, global interceptor, and local interceptor tiers. I also see this module to be an entry point to appeal to Express users that need more extensibility and composability. The `ExpressApplication` and `ExpressServer` shows some of the possibilities.

2. Introduced a `InvokeMiddleware` action to the sequence as the extension point to discover registered middleware and invoke them as an interceptor chain.

3. Moved existing `cors` and `apiSpec` Express middleware into the `InvokeMiddleware` chain. Aligned `mountExpressRouter` with the middleware plumbing.

4. Updated todo example and code generation templates to use the new features.

bajtos commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,490 @@
+
+- As a `Middleware` to be executed by `InvokeMiddleware` actions within the
+  sequence. Middleware are registered by extension points to form chains
+  (`MiddlewareChain').
```
```suggestion
  (`MiddlewareChain`).
```

bajtos commented:
In my view, this proposal is introducing a lot of complexity and many new moving parts. I am not convinced at all that we need this.

IMO, the following tool would be good enough for 80% of LoopBack users.

```ts
// usage
const compression = require('compression');
const helmet = require('helmet');

const middleware = [
  compression(),
  helmet.hidePoweredBy(),
  helmet.ieNoOpen(),
  helmet.noSniff(),
];

export class MySequence implements SequenceHandler {
  constructor(
    // inject dependencies (skipped for brevity)
  ) {}

  async handle(context: RequestContext) {
    try {
      const {request, response} = context;
      if (await executeMiddlewareChain(context, middleware)) {
        // one of the middleware handled the request
        return;
      }
      const route = this.findRoute(request);
      const args = await this.parseParams(request, route);
      const result = await this.invoke(route, args);
      this.send(response, result);
    } catch (error) {
      this.reject(context, error);
    }
  }
}

// implementation (mock-up)
async function executeMiddlewareChain({request, response}, chain) {
  const responseWritten = onFinishedAsync(response).then(() => true);
  const handlerFinished = new Promise<boolean>((resolve, reject) => {
    let ix = 0;
    next();

    function next(err) {
      if (err) return reject(err);
      if (ix <= chain.length) {
        // Express router/middleware called next, which means the request was not handled (yet)
        return resolve(false);
      }
      const hander = chain[ix++];
      handler(request, response, next);
    }
  });
  return Promise.race([handlerFinished, responseWritten]);
}
```

I see the following benefits of my solution:
1. The API is very easy to learn for people that are already familiar with Express.
2. Definition of which middleware to execute stays in the same place where we define other parts of request handling workflow (i.e. the sequence). As I see it, a middleware is just another sequence step.
3. It's just few lines of code, easy to debug when something goes wrong, easy to maintain for us.

>  Express middleware are one-way handlers that mostly process requests and there is no first-class facility to handle responses. We would love to support [Koa style middleware](https://github.com/koajs/koa/blob/master/docs/guide.md#writing-middleware) that use `async/await` to allow cascading behaviors.

I am not sure if we really want that. I believe we have decided very early on that we don't want to use the concept of a middleware and want to use Sequence and Sequence Actions instead. Unfortunately, we are not keeping documentation on the architecture decisions we made. There may be some reasoning captured in discussions in the early pull request and/or issues, IDK.

> Express middleware are added by the order of `app.use()` and it's hard to contribute middleware from other modules.

Based on my experience with LoopBack 2/3 and middleware phases, I don't think this is really an issue. I am not aware of any extension using the middleware phases feature, which was specifically introduced in LB2/3 to allow extensions to do this. On the opposite, I heard from users that they find middleware phases unnecessarily complicated and don't see any benefits.

I see two problems to solve:
1. How to leverage the vast ecosystem of existing Express middleware when building LB4 apps.
2. How to allow LB4 extensions to contribute middleware-like functionality.

I don't think it's a good idea to aim for a single solution to both of them.

1. As you have correctly pointed out, the design of Express middleware has its limitation. At the same time, it's powerful enough to serve millions of Express users. Instead of trying to shoehorn Express middleware onto LB4 programming model, I think it's best to accept the limitations, provide a thin bridge allowing Express middleware to be used in LB4 apps (following the Express way) and advise the users to use (or write) LB4-native extensions once they outgrow the design of Express middleware.

2. Once we get to the point of writing an LB4 extension, we should use LB4 building blocks and design language. An extension wishing to contribute a new sequence action should do exactly that - contribute a sequence action, not a new middleware. I believe we have already validated this approach in several extensions we provide by now (authentication, authorization, logging, etc.)

I don't see a place for a generic Koa-like middleware here.


> Express does not allow dependency injection. We would like to enable injection of middleware configurations.

I don't understand what is the benefit of using DI for middleware configuration. Can you please explain what use cases you have in mind?

To me, this feels like _"if you have a hammer, everything looks like a nail"_.

bajtos commented:
Setting my reservations aside, here are few comments for the current implementation.

(0) Most importantly: let's make sure we make it easy to solve easy problems and not too much difficult to solve more complex ones. Adding a global middleware to a LB4 app should be as close to the simplicity of Express' `app.use(compression())` as possible.

(1) I see this PR is adding new packages, did you follow all steps from [Adding a new package](https://github.com/strongloop/loopback-next/blob/master/docs/site/DEVELOPING.md#adding-a-new-package) checklist?

(2) How do you handle the case when Express middleware does not call `next`? IMO, we must return control back to the sequence class, I don't think your implementation does that. Consider the example from https://github.com/strongloop/loopback-next/pull/2434

```ts
export class MySequence implements SequenceHandler {
  // constructor with @inject decorators

  async handle(context: RequestContext) {
    const {request, response} = context;
    const start = process.hrtime();
    try {
      await this.invokeMiddleware(); 
      const route = this.findRoute(request);
      const args = await this.parseParams(request, route);
      const result = await this.invoke(route, args);
      this.send(response, result);
    } catch (error) {
      this.reject(context, error);
    } finally {
      const delta = process.hrtime(start);
      console.log(`Handled "${request.method} ${request.url}" in ${delta[0]} seconds.`);
    }
  }
}
```

Let's say we have `cors` or `favicon` middleware configured and the middleware handle the request (e.g. returned a response for `OPTIONS` request, returned the favicon image). As a LB4 developer, I would expect a timing log printed for these requests too. I don't think that will happen with your current design?

(3) I find the proposed documentation as too involved for new LB developer. Consider the following user story:

- As a LB4 app developer, I want to add few Express middleware to improve my LB4 application.
- I want `favicon` middleware to be invoked as the first one, _before_ CORS is executed.
- I want to invoke various `helmet` middleware  _after_ CORS is executed, e.g. `helmet.hidePoweredBy()`.
- Middleware ordering may look somewhat arbitrary, but I have intentionally crafted it to force you to consider ordering of middleware in relation to built-in middleware like CORS.

To learn how to do that, I have to read 300+ lines of the docs until I find the line https://github.com/strongloop/loopback-next/pull/5118/files#diff-a429ccda8ffe42b31d5816e9d22a27b5R325.

I think it would be better to structure the docs as follows:

1. Start with a step-by-step guide showing how to solve the most common use case (add a new middleware to the sequence at a given position relative to other middleware). Keep this high-level - describe how to achieve the outcome, don't explain how the sausage is made. Most LB4 users should not need to learn more than this.

2. Then provide content for users that need more than just the basic happy path. Here you can explain the concept of Express, Koa and LoopBack middleware, the types and implementation patterns involved, etc. I think your current docs content is pretty much covering this part :+1:

bajtos commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,490 @@
+  IHelmetConfiguration
+> {
+  constructor(
+    @config()
```
+1 for using existing `@config` infrastructure 👏

jannyHou commented on packages/express/src/__tests__/acceptance/test-helpers.ts:
```diff
@@ -0,0 +1,117 @@
+    this.app.expressServer.expressApp.post('/hello', async (req, res, next) => {
+      try {
+        const controller = await this.app.get<MyController>(binding.key);
+        const proxy = createProxyWithInterceptors(
```
hmm why `proxy` is needed here?

victor-enogwe commented:
@raymondfeng nice work on your pull request, I recently started using loopback, great framework, thanks to all the contributors. I do hope this pull-request solves most of the middleware issues. 

Here is my experience so far with adding an express middleware.

I wanted to run a session middleware before the decorated method using an interceptor. That wasn't happening as according to the [docs](https://loopback.io/doc/en/lb4/Interceptors.html#apply-interceptors), the `@interceptor` decorator only adds metadata to the method and does not wrap the method. 

This metadata is then used to determine when the interceptor will be called based on the grouping. I did not find in the documentation how to make these interceptors run before my `@authenticate` decorated route method. 

To solve this I resulted in either calling the middleware in the sequence handler or adding them directly to my strategy classes or to my controller classes. That sort of introduces an inconsistent way of injecting middleware into the app, which would have been really elegant using only `interceptors`.

I think interceptors already do a great job with adding middleware, and probably just needs a way to determine where and when to call the middleware method, plus loopback does a great job at making almost any app context globally available.

What do you think of the `Hooks and Anchors pattern`(I think that's the name), with priority/grouping as implemented already to add middleware and also determine where the middleware will be called? 

An example would be WordPress's use of hooks like `add_action` and `do_action` to specify where and when to run a function.

raymondfeng commented:
@bajtos Thank you for the feedback and opinions. I have incorporated some of them into the latest code, such as:

1. Allow `InvokeMiddleware` to run an explicit list of Express middleware handler functions
2. Add more flavors of `expressMiddleware()` to register Express middleware handler/factory functions

I also refactored the `Express-middleware.md` into two files:

- Middleware.md -Introduction of Middleware and the architectural rationale   
- Express-middleware.md - Serving as a user guide to plug in Express middleware into LoopBack as part of sequence actions and interceptors.

I would like to justify why I want to open up `Sequence` to allow pluggable middleware.

1. The `Middleware` and `MiddlewareChain` are not new concepts. They are just specialization of `GenericInterceptor` and `GenericInterceptorChain`.  We already use the same concept for method invocations. Being able to use the same programming model concepts for different layers provide the consistency and ease of learning. I also envision to have similar chains to intercept outbound interactions for datasources/connectors.
 
2. The core of LoopBack is built upon the ideas of IoC, DI, and Extension Point/Extension. We can achieve IoC for most parts of LoopBack except `Sequence`, which is hard-coded actions generated by CLI. It's not very friendly to contribute actions from extension modules using `Component` today. Let me show a few examples:

    - `@loopback/authentication` cannot be plugged into an application by `this.component(AuthenticationComponent)`. The manual step to add `Authenticate` action violates the design of `Component` to some extent. 
    - In contrast, `@loopback/authorization` contributes a global interceptor without requiring extra steps to set up the application beyond mounting the component.   
   - To contribute a `Helmet` component for LoopBack, I would expect it should be possible to contribute middleware without requiring manual update of `Sequence`. The prospect component should be able to leverage DI to customize the configuration.

3. I would like to allow metadata driven composition of the sequence. Imagine a LoopBack `starter` can simply bundle itself as a component that sets up bindings as its template to customize what features to be used and how they are configured.  I don't understand why `Sequence` has to be hard-coded.  For debugging, we have good trace of interceptor chains. It's not different from contributing interceptors.

4. I would like to allow configuration to be updated at runtime by simply rebinding the configuration value and make it possible for the middleware to pick up. For example, adjust logging level from `info` to `debug` if necessary.

5. I would to follow a simple principle: Make it easy to do simple things while make it possible to do advanced things. I agree the PR can solve two of the related problems but it's shooting two birds in one stone, IMO. :-).
    - How to leverage the vast ecosystem of existing Express middleware when building LB4 apps.
    - How to allow LB4 extensions to contribute middleware-like functionality.
We're not reinventing the wheel as we are just extending the interceptor idea to new use cases. We also want to allow adapting and applying Express middleware as invocation interceptors anyway.

Maybe we can find a good middle ground to move forward quickly as follows:

1. I'll try to make APIs as simple as possible for your use cases from whom just wants to drop in Express middleware - similar as `app.use`, for example `RestServer.use`?

2. I have added acceptance tests in the PR to demonstrate some of the advanced use cases I have in mind. You can propose new ideas.

3. We don't have to be perfect for the first iteration. We can always improve the design and refine it over time.

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+## Overview
+
+Express is the most popular web framework for Node.js developers. As quoted
+below from
```
> below from

below from the

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+## Bring Express middleware to LoopBack
+
+Let's start with a few examples to illustrate how we can bring Express
+middleware to LoopBack applications with minimal efforts.
```
> efforts

effort

raymondfeng replied:
Google search shows `minimal efforts` is more commonly used.

emonddr replied:
hmm. doesn't sound right. See https://dictionary.cambridge.org/example/english/minimal-effort

raymondfeng replied:
I accept `minimal effort` 👍

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+### Use Express middleware within the sequence of actions
+
+Express middleware can now be plugged into the REST sequence with an
+`InvokeMiddleware` action being injected to the default sequence.
```
> Express middleware can now be plugged into the REST sequence with an `InvokeMiddleware` action being injected to the default sequence.

Express middleware can now be plugged into the default REST sequence via the `InvokeMiddleware` action.

emonddr replied:
Is this section the same as `#### Register Express middleware to be executed by `InvokeMiddleware` actions`?

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+        helmet({}), // options for helmet is fixed and cannot be changed at runtime
+        morgan('combined', {}), // options for morgan is fixed and cannot be changed at runtime
+      ]);
+      if (finished) {
```
What is `finished`'s value exactly? What if I want to perform some middleware, and then continue with the sequence actions below?

raymondfeng replied:
`finished` indicates that http response is produced by one of the middleware and we should not proceed with other actions.

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+
+### Use Express middleware as interceptors for controllers
+
+The LoopBack global and local interceptors now also serve as a revenue to attach
```
what you mean by `revenue ` ?

raymondfeng replied:
Ha, it should be `avenue`.

emonddr commented on docs/site/Express-middleware.md:
````diff
@@ -0,0 +1,317 @@
+```
+
+If the Express middleware module does not expose a factory function conforming
+to `ExpressMiddlewareFactory` signature, a wrapper can be created. For example:
````
> to `ExpressMiddlewareFactory` signature

to the `ExpressMiddlewareFactory` signature

emonddr commented on docs/site/Express-middleware.md:
````diff
@@ -0,0 +1,317 @@
+const morganFactory = (config?: morgan.Options) => morgan('combined', config);
+```
+
+### Define a provider class for middleware based interceptor
````
> middleware based 

middleware-based

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+
+It's often desirable to allow dependency injection of middleware configuration
+for the middleware. We can use `defineInterceptorProvider` to simplify
+definition of such provider classes.
```
> definition of

the definition of

emonddr commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,317 @@
+
+### Use `lb4 interceptor` command to create interceptors for Express middleware
+
+The `lb4 interceptor` can be used to generate skeleton implementation of global
```
> generate skeleton implementation

generate a skeleton implementation for global or local

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,282 @@
+![middleware](imgs/middleware.png)
+
+In the diagram, a specialized type of interceptors is introduced as `Middleware`
+for LoopBack. Our middleware is basically a function with the signature:
```
the signature is missing here.

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,282 @@
+
+#### Default sequence
+
+The generated `src/sequence.ts` now has a new `middleware` property to
```
> to `DefaultSequence`

in `DefaultSequence`

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,282 @@
+#### Default sequence
+
+The generated `src/sequence.ts` now has a new `middleware` property to
+`DefaultSequence` to allow injection of the middleware chain while keeping
```
> while keeping

while keeping the

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,282 @@
+    try {
+      const {request, response} = context;
+      // The default middleware chain
+      await this.invokeMiddleware(context);
```
no check for `finished` anymore on top and bottom?

emonddr replied:
why not specifiy :  PRE_INVOCATION_MIDDLEWARE ?

raymondfeng replied:
That's the default.

emonddr commented on docs/site/Middleware.md:
````diff
@@ -0,0 +1,282 @@
+```
+
+It's also possible to inject multiple instances of `InvokeMiddleware`, each of
+which has its own extension point name.
````
What is the `different` extension point name being used below? It is not obvious.

emonddr commented on docs/site/Middleware.md:
````diff
@@ -0,0 +1,282 @@
+```ts
+export class MyApplication extends RestApplication {
+  constructor(config: ApplicationConfig) {
+    // Create another middleware phase. This is NOT needed if we use the same
````
Not sure what the comment means. Can you clarify?  we seem to be using invokeMiddleware WITH a different extension, so the comment is confusing.

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,282 @@
+    // Register a middleware
+    // `spy` is the factory function of a middleware
+    //
+    const spyBinding = this.expressMiddleware(spy, undefined, {
```
what is the undefined for (spy,undefined) ?

bajtos commented:
> I would like to justify why I want to open up `Sequence` to allow pluggable middleware.

Thank you for providing more details behind your decision. Can we capture this somewhere in our docs please for future readers? (See [Lightweight Architecture Decision Records](https://www.thoughtworks.com/radar/techniques/lightweight-architecture-decision-records).) Maybe in `Middleware.md`?

>  I would like to allow metadata driven composition of the sequence. Imagine a LoopBack `starter` can simply bundle itself as a component that sets up bindings as its template to customize what features to be used and how they are configured.  I don't understand why `Sequence` has to be hard-coded.  For debugging, we have good trace of interceptor chains. It's not different from contributing interceptors.

Fair enough. Maybe we have outgrown the ideas behind Sequence design and it's time to abandon Sequence in favor of something else?

From what I remember, @ritch introduced Sequence as a way how to avoid all the different problems we observed with Express middleware in LoopBack 3. I hope that by moving from Sequence to Middleware/Interceptor chain, we won't reintroduce those problem back again.


> Maybe we can find a good middle ground to move forward quickly as follows:
> 
> 1. I'll try to make APIs as simple as possible for your use cases from whom just wants to drop in Express middleware - similar as `app.use`, for example `RestServer.use`?

Sure. I am not sure if `use` is the best name as it's not clear that it's for middleware, but that's a detail.

> 2. I have added acceptance tests in the PR to demonstrate some of the advanced use cases I have in mind. You can propose new ideas.

Cool, I'll take a look.

> 3. We don't have to be perfect for the first iteration. We can always improve the design and refine it over time.

In my experience, some design aspects are difficult to change in a backwards compatible way. The difficult part is to tell which aspects are easy to refine and which not 🤔

bajtos commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,319 @@
+      // Express middleware handler functions
+      const finished = await this.invokeMiddleware(context, [
+        helmet({}), // options for helmet is fixed and cannot be changed at runtime
+        morgan('combined', {}), // options for morgan is fixed and cannot be changed at runtime
```
Please note we must not call middleware factories for every incoming request, that would be very bad for performance.

Proposed change:

```ts
const middleware = [
  helmet({}), // options for helmet is fixed and cannot be changed at runtime
  morgan('combined', {}), // options for morgan is fixed and cannot be changed at runtime
];

export class MySequence extends DefaultSequence {
   async handle(context: RequestContext): Promise<void> {
      // ...
      const finished = await this.invokeMiddleware(context, middleware);
      // ...
  }
}
```

bajtos commented on docs/site/Express-middleware.md:
```diff
@@ -0,0 +1,319 @@
+
+export class MyApplication extends RestApplication {
+  constructor(config: ApplicationConfig) {
+    this.expressMiddleware(
```
Maybe `useExpressMiddleware` or `useMiddleware` would be a better name? The usual convention for method names is to start with a verb.

raymondfeng replied:
I choose to use `expressMiddleware` to be consistent with other registration related apis, such as:

- app.controller()
- app.component()
- app.dataSource()
- app.service()

bajtos commented on docs/site/Express-middleware.md:
````diff
@@ -0,0 +1,319 @@
+```ts
+import helmet, {IHelmetConfiguration} from 'helmet';
+const binding = registerExpressMiddlewareInterceptor(
+  app,
````
Where is `app` coming from? Where should be this code snippet placed in a LB4 application?

bajtos commented on docs/site/sidebars/lb4_sidebar.yml:
```diff
@@ -103,6 +103,14 @@ children:
 
+  - title: 'Middleware'
+    url: Middleware.html
+    output: 'web, pdf'
```
I feel this page does not belong to [Key Concepts](https://loopback.io/doc/en/lb4/Concepts.html) and should be moved to [Behind the scene](https://loopback.io/doc/en/lb4/Behind-the-scene.html).

I am expecting that new LB users won't need to understand the information from this page until they get into really advanced use cases.

raymondfeng replied:
I put it under `Key Concepts` for two reasons:

- It's next to `Interceptors`
- Application developers need to add middleware

bajtos commented on docs/site/sidebars/lb4_sidebar.yml:
```diff
@@ -103,6 +103,14 @@ children:
+
+  - title: 'Using Express Middleware'
+    url: Express-middleware.html
+    output: 'web, pdf'
```
I think this page should be nested inside [Sequence](https://loopback.io/doc/en/lb4/Sequence.html). Please note that the Sequence page already provides some content related to Express-middleware, see [Working with Express middleware](https://loopback.io/doc/en/lb4/Sequence.html#working-with-express-middleware).

Please update that existing content to work well with your new pages and to avoid duplicative and/or contradicting information.

_I have also briefly considered the option to put this page into [Usage scenarios](https://loopback.io/doc/en/lb4/Usage-scenarios.html), but since we already have some middleware-related content in Sequence, it may be best to put the new content there too._

bajtos replied:
Also, do you have any plans for https://loopback.io/doc/en/lb4/express-with-lb4-rest-tutorial.html? Should we update it to mention the newer (and better) way for mounting Express middleware? Perhaps we should deprecate and eventually remove that page?

raymondfeng replied:
> I think this page should be nested inside Sequence. Please note that the Sequence page already provides some content related to Express-middleware, see Working with Express middleware.

We might have to use this page for two cases:

- Apply Express middleware in the sequence
- Apply Express middleware as interceptors

bajtos commented on packages/express/README.md:
```diff
@@ -0,0 +1,153 @@
+This package adds middleware support for LoopBack 4 and allows Express
+middleware to be plugged into LoopBack seamlessly. It's used by `@loopback/rest`
+to support Express middleware with `InvokeMiddleware` action within the REST
+sequence.
```
Is it worth the effort to maintain this feature as a standalone package, when it's not an optional extension and it's always used by (included in) `@loopback/rest`?

raymondfeng replied:
I think so. My plan is to allow the usage of an existing Express application without dragging in `@loopback/rest`.

bajtos commented on packages/express/package.json:
```diff
@@ -0,0 +1,65 @@
+  ],
+  "dependencies": {
+    "@loopback/context": "^3.5.1",
+    "@loopback/core": "^2.4.1",
```
Please make sure this package is using the version of `@loopback/core` and `@loopback/context` as provided by the target LB4 application, so that app users are in full control.

Proposed changes:
- Move `@loopback/core` to peer dependencies
- Remove `@loopback/context` entirely, import context-related APIs from `@loopback/context`.

raymondfeng replied:
We can remove `@loopback/context` and use `@loopback/core` only. But we need to apply the same rule for all other packages.

I'm also reluctant to use use `peerDependencies`. IMO, the only reliable way is to use the same set of versions released together and upgrade using `lb4 update`.

Let's defer this to a later time that we can visit all packages in `loopback-next`.

bajtos commented on packages/rest/src/rest.server.ts:
```diff
@@ -251,7 +242,11 @@ export class RestServer extends Context implements Server, HttpServerLike {
+      injectConfiguration: false,
+      key: 'middleware.cors',
+      group: 'cors',
+    });
```
I think we should eventually not rely on the built-in CORS middleware and modify our application template to include explicit setup of CORS.

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,335 @@
+      const result = await this.invoke(route, args);
+
+      // The second middleware chain for post-invocation processing
+      context.bind('invocation.result').to(result);
```
What does this do? Is it needed?

raymondfeng replied:
This shows the possibility to access results from other actions inside a middleware.

raymondfeng replied:
Added more comments.

emonddr commented on docs/site/Middleware.md:
```diff
@@ -0,0 +1,335 @@
+      const result = await this.invoke(route, args);
+
+      // The second middleware chain for post-invocation processing
+      context.bind('invocation.result').to(result);
```
Why is this needed?

emonddr commented on packages/rest/src/__tests__/acceptance/middleware/middleware-sequence.acceptance.ts:
```diff
@@ -0,0 +1,177 @@
+        const result = await this.invoke(route, args);
+
+        // The second middleware chain for post-invocation processing
+        context.bind('invocation.result').to(result);
```
Why is this needed?

emonddr commented on packages/rest/src/__tests__/acceptance/middleware/middleware-sequence.acceptance.ts:
```diff
@@ -0,0 +1,177 @@
+        const result = await this.invoke(route, args);
+
+        // The second middleware chain for post-invocation processing
+        context.bind('invocation.result').to(result);
```
Why is this needed?

emonddr commented on docs/site/Middleware.md:
````diff
@@ -0,0 +1,335 @@
+```ts
+import {ApplicationConfig} from '@loopback/core';
+import {RestApplication} from '@loopback/rest';
+import {log} from './log-middleware';
````
Where is `log-middleware` defined?

raymondfeng replied:
I added a caption for 1st reference.

jannyHou commented on docs/site/Express-middleware.md:
````diff
@@ -0,0 +1,327 @@
+#### Create an interceptor from Express middleware factory function
+
+```ts
+import {createInterceptor} from '@loopback/express';
````
could you explain more about when use `createInterceptor` and when use `toInterceptor`?
e.g. for this example, why not using:

`const helmetInterceptor = toInterceptor(helmet(helmetConfig));`

raymondfeng replied:
`toInterceptor()` takes the Express handler function directly while `createInterceptor` takes the factory function and config object.

jannyHou commented on packages/express/src/__tests__/acceptance/test-helpers.ts:
```diff
@@ -0,0 +1,129 @@
+      }
+    }
+    const binding = this.app.controller(MyController);
+    this.app.expressServer.expressApp.post('/hello', async (req, res, next) => {
```
a bit confused: the testing endpoint is set up by express server, why not use the REST decorators? Is there a difference?

raymondfeng replied:
This module does not have dependency on `@loopback/rest` by design. It should possible to use it with plain Express.

deepakrkris replied:
Any advantage we have, by this being an independent package ?

raymondfeng commented on docs/site/Express-middleware.md:
````diff
@@ -0,0 +1,321 @@
+```
+
+If the Express middleware module does not expose a factory function conforming
+to the `ExpressMiddlewareFactory` signature, a wrapper can be created. For
````
Yes

jannyHou commented:
> @loopback/authentication cannot be plugged into an application by this.component(AuthenticationComponent). The manual step to add Authenticate action violates the design of Component to some extent.
    In contrast, @loopback/authorization contributes a global interceptor without requiring extra steps to set up the application beyond mounting the component.
    To contribute a Helmet component for LoopBack, I would expect it should be possible to contribute middleware without requiring manual update of Sequence. The prospect component should be able to leverage DI to customize the configuration.

👍 👍 After this PR we can register the anthenticate action into a middleware, then no need to manually modify the sequence.

Nice docs and feature! A few questions.

raymondfeng commented on packages/express/src/express.server.ts:
```diff
@@ -0,0 +1,118 @@
+/**
+ * An Express server that provides middleware composition and injection
+ */
+export class ExpressServer extends MiddlewareRegistry implements Server {
```
We could make `MiddlewareRegistry` a mixin if needed.

raymondfeng commented on packages/rest/src/sequence.ts:
```diff
@@ -95,6 +103,8 @@ export class DefaultSequence implements SequenceHandler {
     try {
       const {request, response} = context;
+      // Invoke registered Express middleware
+      await this.invokeMiddleware(context);
```
Yes.
